### PR TITLE
Add resuming paging

### LINKED ISSUE #219: Resume paging
In cdrs(_tokio), the user can page with state: https://github.com/krojew/cdrs-tokio/blob/40110835db2f3695e8176a8250caa67ca22e22bc/examples/paged_query.rs#L210. This means the server can send clients the current paging state, which in turn can be used to retrieve the next page. 

I don't see any option how to set/get the current cursor for paging. I also can not see how to specify the number of rows to retrieve for each page. I would expect to see this functionality demonstrated in https://github.com/scylladb/scylla-rust-driver/blob/main/examples/select-paging.rs.

### PR DESCRIPTION
**Pull request by psarna**
This draft adds a way of resuming queries from previously received paging state. The paging state can be extracted from query results and passed to new queries to let them resume execution from where the previous query left off. It's an alternative way of querying to the existing `query_iter` mechanism.

This series is marked as a draft because of the following TODO list:
 - [x] Add tests for CI.
 - [x] Add documentation.
 - [ ] <strike>Currently the paging state is copied underneath. Consider optimizing this copying out - it's not likely to be a performance killer because it is expected to fit into a single cacheline, but eliding copies is good practice anyway. That would likely require some changes in function prototypes. Perhaps a good candidate for a follow-up patch.</strike>

Fixes #219

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Comment by piodul** [comment]
> Currently the paging state is copied underneath. Consider optimizing this copying out - it's not likely to be a performance killer because it is expected to fit into a single cacheline, but eliding copies is good practice anyway. That would likely require some changes in function prototypes. Perhaps a good candidate for a follow-up patch.

[`Bytes`](https://docs.rs/bytes/1.0.1/bytes/struct.Bytes.html) is cheaply cloneable (it's like a shared atomic pointer to a slice), so no need to need to worry about avoiding allocations. However, the `Bytes` structure itself is quite large - it's 4 `usize`s. I'm not sure how large the paging state can be, but if it is smaller we may consider switching to a smaller shared pointer type, e.g. [`Arc`](https://doc.rust-lang.org/std/sync/struct.Arc.html).

**Comment by psarna** [comment]
> > Currently the paging state is copied underneath. Consider optimizing this copying out - it's not likely to be a performance killer because it is expected to fit into a single cacheline, but eliding copies is good practice anyway. That would likely require some changes in function prototypes. Perhaps a good candidate for a follow-up patch.
> 
> [`Bytes`](https://docs.rs/bytes/1.0.1/bytes/struct.Bytes.html) is cheaply cloneable (it's like a shared atomic pointer to a slice), so no need to need to worry about avoiding allocations. However, the `Bytes` structure itself is quite large - it's 4 `usize`s. I'm not sure how large the paging state can be, but if it is smaller we may consider switching to a smaller shared pointer type, e.g. [`Arc`](https://doc.rust-lang.org/std/sync/struct.Arc.html).

Good to know about Bytes. And paging state is easily larger than that - among other things, it encodes the primary key, and that alone can be several kilobytes in extreme cases.

File: scylla/src/transport/session.rs [hunk 1]
```diff
@@ -409,9 +410,25 @@ impl Session {
...
+        self.execute_paged(prepared, values, None).await
+    }
+
+    /// Executes a previously prepared statement with previously received paging state
+    /// # Arguments
+    ///
+    /// * `prepared` - a statement prepared with [prepare](crate::transport::session::Session::prepare)
+    /// * `values` - values bound to the query
+    /// * `paging_state` - paging state from the previous query or None
+    pub async fn execute_paged(
```
Comment (piodul) [inline scylla/src/transport/session.rs:438]: Have you considered adding a `paging_state` field to `Query` and `PreparedStatement`? My main worry here is that if we decide to add another optional parameter like that, you will need to have a separate method for each combination of "with parameter"/"without parameter", or force the user to always provided `None` or something like that.

On the other hand, the problem with `paging_state` field is that somebody might forget that they set it...

File: scylla/src/statement/query.rs [hunk 2]
```diff
@@ -28,6 +28,18 @@ impl Query {
...
+        Self {
+            contents,
+            page_size: Some(page_size),
+            consistency: Default::default(),
+            serial_consistency: None,
+            is_idempotent: false,
+            retry_policy: None,
+            tracing: false,
+        }
+    }
```
Comment (piodul) [inline scylla/src/statement/query.rs:41]: You can make this function much shorter:

```rust
pub fn with_page_size(contents: String, page_size: i32) -> Self {
    Self {
        page_size: Some(page_size),
        ..Self::new(contents)
    }
}
```

Comment (piodul) [reply scylla/src/statement/query.rs:41]: On the other hand, maybe we could convert it to a builder-style method? It would take `self` by value and return a new query:

```rust
pub fn with_page_size(mut self, page_size: i32) -> Self {
    self.page_size = page_size;
    self
}
```

Usage:

```rust
// Before:
Query::with_page_size(contents, page_size)

// After:
Query::new(contents).with_page_size(page_size)
```

Slightly longer, but with this pattern you can add more `with_xyz` methods and chain them.

File: scylla/src/statement/prepared_statement.rs [hunk 3]
```diff
@@ -22,13 +22,18 @@ pub struct PreparedStatement {
 }
 
 impl PreparedStatement {
-    pub fn new(id: Bytes, metadata: PreparedMetadata, statement: String) -> Self {
+    pub fn new(
+        id: Bytes,
+        metadata: PreparedMetadata,
+        statement: String,
+        page_size: Option<i32>,
+    ) -> Self {
```
Comment (piodul) [inline scylla/src/statement/prepared_statement.rs:30]: Instead of an `Option` parameter, maybe we could add a builder-like `with_page_size` method, too?

File: examples/select-paging.rs [hunk 4]
```diff
@@ -40,22 +40,58 @@ async fn main() -> Result<()> {
...
+    );
+    let res6 = session
+        .execute_paged(&paged_prepared, &[], res5.paging_state)
+        .await?;
+    println!(
+        "Paging state from the third prepared statement execution: {:#?} ({} rows)",
+        res6.paging_state,
+        res6.rows.unwrap().len()
     );
-
```
Comment (piodul) [inline examples/select-paging.rs]: Unnecessary whitespace change?

File: scylla/src/transport/session.rs [hunk 5]
```diff
@@ -322,13 +338,12 @@ impl Session {
...
             |connection: Arc<Connection>| async move {
                 connection
-                    .query_single_page_by_ref(query_ref, values_ref)
+                    .query_single_page_by_ref(query_ref, values_ref, paging_state_ref)
                     .await
             },
         )
         .await
     }
-
```
Comment (piodul) [inline scylla/src/transport/session.rs:330]: Unnecessary whitespace change?

File: scylla/src/transport/session.rs [hunk 6]
```diff
@@ -449,7 +464,9 @@ impl Session {
             retry_session,
             |node: Arc<Node>| async move { node.connection_for_token(token).await },
             |connection: Arc<Connection>| async move {
-                connection.execute_single_page(prepared, values_ref, paging_state_ref).await
+                connection
+                    .execute_single_page(prepared, values_ref, paging_state_ref)
+                    .await
```
Comment (piodul) [inline scylla/src/transport/session.rs:469]: This formatting change probably belongs to the previous commit.

File: scylla/src/transport/connection.rs [hunk 7]
```diff
@@ -312,7 +313,7 @@ impl Connection {
         &self,
         prepared_statement: &PreparedStatement,
         values: impl ValueList,
-        paging_state: Option<Bytes>,
+        paging_state: &Option<Bytes>,
```
[the remark this answers is not in the document]
Comment (psarna) [reply scylla/src/transport/connection.rs:319]: I wish I could, but it's called from within a lambda which implements Fn (and not FnOnce), and making paging_state a reference here was the least painful way of dealing with the borrow checker I thought of (also, same trick is applied to values nearby). Suggestions welcome :)

File: scylla/src/transport/session.rs (hunk 1 above)
Comment (psarna) [reply scylla/src/transport/session.rs:438]: I did, but I decided it doesn't belong to any particular query or statement. It's just a way of resuming a previous run of a particular query, so this value is expected to change on each call to execute()/query()

File: scylla/src/transport/connection.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/connection.rs:319]: Here you go:

```diff
diff --git a/scylla/src/transport/connection.rs b/scylla/src/transport/connection.rs
index f256b3a..a14ca0a 100644
--- a/scylla/src/transport/connection.rs
+++ b/scylla/src/transport/connection.rs
@@ -305,7 +305,7 @@ impl Connection {
         &self,
         prepared_statement: &PreparedStatement,
         values: impl ValueList,
-        paging_state: &Option<Bytes>,
+        paging_state: Option<Bytes>,
     ) -> Result<QueryResult, QueryError> {
         self.execute(prepared_statement, values, paging_state)
             .await?
@@ -316,7 +316,7 @@ impl Connection {
         &self,
         prepared_statement: &PreparedStatement,
         values: impl ValueList,
-        paging_state: &Option<Bytes>,
+        paging_state: Option<Bytes>,
     ) -> Result<QueryResponse, QueryError> {
         let serialized_values = values.serialized()?;
 
@@ -327,7 +327,7 @@ impl Connection {
                 serial_consistency: prepared_statement.get_serial_consistency(),
                 values: &serialized_values,
                 page_size: prepared_statement.get_page_size(),
-                paging_state: paging_state.clone(),
+                paging_state,
             },
         };
 
diff --git a/scylla/src/transport/iterator.rs b/scylla/src/transport/iterator.rs
index ac84a42..4de0805 100644
--- a/scylla/src/transport/iterator.rs
+++ b/scylla/src/transport/iterator.rs
@@ -159,7 +159,7 @@ impl RowIterator {
 
             let page_query = |connection: Arc<Connection>, paging_state: Option<Bytes>| async move {
                 connection
-                    .execute(prepared_ref, values_ref, &paging_state)
+                    .execute(prepared_ref, values_ref, paging_state)
                     .await
             };
 
diff --git a/scylla/src/transport/session.rs b/scylla/src/transport/session.rs
index 6f58758..7b0189a 100644
--- a/scylla/src/transport/session.rs
+++ b/scylla/src/transport/session.rs
@@ -465,7 +465,7 @@ impl Session {
             |node: Arc<Node>| async move { node.connection_for_token(token).await },
             |connection: Arc<Connection>| async move {
                 connection
-                    .execute_single_page(prepared, values_ref, paging_state_ref)
+                    .execute_single_page(prepared, values_ref, paging_state_ref.clone())
                     .await
             },
         )
```

I think the lambda you wrote about is at the bottom of the diff - explicitly cloning `paging_state_ref` does the trick. In case of the lambda, cloning is necessary because it can be called multiple times, so this diff does not change the number of clones made there.

On the other hand, look at the diff in `RowIterator` - in that case there will be one less clone because the paging state will be passed by value and `execute` won't have to clone it.

File: scylla/src/transport/session.rs (hunk 1 above)
Comment (piodul) [reply scylla/src/transport/session.rs:438]: I see. In that case, I think the current interface is OK.

File: scylla/src/transport/connection.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (psarna) [reply scylla/src/transport/connection.rs:319]: :facepalm: right... I was almost there with paging_state_ref and gave up, while the only thing left to make it work was to `clone()` in lambda's body. Thx, will do in v2

**Comment by psarna** [comment]
v2:
 - applied the comments 
 - added checks to tests

File: scylla/src/transport/session_test.rs [hunk 8]
```diff
@@ -168,7 +187,26 @@ async fn test_prepared_statement() {
         let a = r.columns[0].as_ref().unwrap().as_int().unwrap();
         let b = r.columns[1].as_ref().unwrap().as_int().unwrap();
         let c = r.columns[2].as_ref().unwrap().as_text().unwrap();
-        assert_eq!((a, b, c), (17, 16, &String::from("I'm prepared!!!")))
+        assert_eq!((a, b, c), (17, 16, &String::from("I'm prepared!!!")));
+
+        let mut results_from_manual_paging: Vec<Row> = vec![];
+        let query = Query::new("SELECT a, b, c FROM ks.t2".to_owned()).with_page_size(1);
```
Comment (piodul) [inline scylla/src/transport/session_test.rs:193]: Does Scylla honor such small page sizes? Quote from the protocol spec:

```
  - Clients should not rely on the actual size of the result set returned to
    decide if there are more results to fetch or not. Instead, they should always
    check the Has_more_pages flag (unless they did not enable paging for the query
    obviously). Clients should also not assert that no result will have more than
    <result_page_size> results. While the current implementation always respects
    the exact value of <result_page_size>, we reserve the right to return
    slightly smaller or bigger pages in the future for performance reasons.
```

Comment (psarna) [reply scylla/src/transport/session_test.rs:193]: It does - Scylla will happily return **less** rows due to performance reasons (or even zero), but it will never return more than requested - the rough idea is that if clients experience performance problems due to too small pages, they should bump the page size themselves. Having a page size of 1 makes very little sense in production environments, but it's useful for testing.

**Review by piodul — APPROVED** [review]
